actors/abi/bitfield.go

```diff
@@ -0,0 +1,232 @@
+package abi
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+
+	rlepluslazy "github.com/filecoin-project/specs-actors/actors/util/rleplus"
+)
+
+var ErrBitFieldTooMany = errors.New("to many items in RLE")
+
+type BitField struct {
+	rle rlepluslazy.RLE
+
+	bits map[uint64]struct{}
```

Comment:
I trust this is a direct port from Lotus, but I must admit great confusion from reading the code what the rules or expectations are about when one or the other (or both?) of these fields is actually initialized.

I wish there were tests to come with it demonstrating expected use.

Reply:
In essence, as implemented in lotus. BitField can either be used for listing all bits (`All`, `AllMap`) or adding bits.
When bits are added, they are set in the `bits` map and latter folded into RLE.

If there is a need of it, I can add `Has` function, adding `Delete` function would be harder.

Reply:
It looks like the the contents of `bits` is not preserved after marshaling and unmarshaling, is that correct? 

If that is the case `Has()` logic could (naively) be implemented by calling `AllMap` and checking the map for existence? And `Unset` could be implemented by calling `All`, removing the value from the slice and then passing it to `BitFieldFromSet`?

Reply:
Yes, it could be implemented as such. For removing you would have to change how the marshaling is handled currently it does `or` operation between existing bitset and `bits`.